### DISCUSSION TITLE
Integration with parallel execution frameworks

### DISCUSSION THREAD

**Discussion by marcelned** [duckdb/duckdb#2013] [General]
How does DuckDB integrate with Dask/Ray/Modin?

Are there plans to make it work with such frameworks?

**Comment by Alex-Monahan** [comment]
Which "direction" of integration are you interested in? Running DuckDB queries on Dask/Ray/Modin objects, or exporting DuckDB query results to those libraries? If you are interested in exporting results to them, you can use fetch_df_chunk to pull/load large result sets in chunks so you don't run out of memory. Otherwise, DuckDB benefits quite a lot from sharing memory across threads on a single machine, so it doesn't support spreading queries across a cluster today.
If you want to parallelize across CPU cores on a single machine, check out the threads Pragma!

**Reply by mpage-uga** [reply]
How would you use threads Pragma through python API ?

**Reply by Mytherin** [reply]
You need to create a connection, which allows you to execute arbitrary queries, e.g.:

```sql
import duckdb
con = duckdb.connect();
con.execute("PRAGMA threads=4");
con.execute("...");
```

**Reply by mpage-uga** [reply]
Perfect ! Looking at the answer, it is seems obvious, but I didn't figure it out from the python API doc and examples. Maybe insert an example like this kind in this doc ? BTW, DuckDB doc quality is in general not on par with the wonderful piece of software that you have developed. Kudos !

**Reply by Mytherin** [reply]
You are absolutely correct that the relational API in general is not well documented right now, primarily because it is still under development. We are planning some more changes/overhauls there. Hopefully we will get to this soon!

> Kudos !

Thanks!
